pdr_backend/pdr_dashboard/util/helpers.py

```diff
@@ -90,4 +154,69 @@ def get_date_period_text_for_selected_predictoors(payouts: pl.DataFrame) -> str:
 
 @enforce_types
 def get_date_period_text_header(start_date: UnixTimeS, end_date: UnixTimeS) -> str:
-    return _format_date_text(start_date.to_dt(), end_date.to_dt(), True)
+    return _format_date_text(start_date.to_dt(), end_date.to_dt())
+
+
+def get_empty_feeds_filters() -> List[Filter]:
+    return [Filter(**item) for item in EMPTY_FEEDS_FILTERS]
+
+
+@enforce_types
+def produce_feeds_filter_options(
+    df: pl.DataFrame,
+) -> Tuple[List[str], List[str], List[str], List[str]]:
+    df = df.with_columns(
+        pl.col("pair")
+        .str.split_exact("/", 1)
+        .map_elements(lambda x: x["field_0"], return_dtype=pl.String)
+        .alias("base_token"),
+        pl.col("pair")
+        .str.split_exact("/", 1)
+        .map_elements(lambda x: x["field_1"], return_dtype=pl.String)
+        .alias("quote_token"),
+        pl.col("source").str.to_titlecase().alias("source"),
+    )
+
+    return (
+        df["base_token"].unique().to_list(),
+        df["quote_token"].unique().to_list(),
+        df["source"].unique().to_list(),
+        df["timeframe"].unique().to_list(),
+    )
+
+
+def check_data_loaded(
+    output_count=1,
+    alternative_output: Optional[Union[Dict[str, Any], List[Dict[str, Any]]]] = None,
+):
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            is_initial_data_loaded = args[-1]  # assuming the last argument is the state
+            if not is_initial_data_loaded or not is_initial_data_loaded["loaded"]:
+                if alternative_output:
+                    return alternative_output
+
+                if output_count > 1:
+                    return tuple(
+                        dash.no_update for _ in range(output_count)
+                    )  # multiple outputs
+
+                return dash.no_update  # single output
+            return func(*args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+def get_feeds_for_favourite_predictoors(app, feed_data, predictoor_addrs):
```

Comment:
I think this can be in the data manager class and replace all app.data with self